File: Lecture 11/Python Code/Tomato/models/Order.py
class Order:
    next_order_id = 1

    def __init__(self, user, restaurant, items, order_type, scheduled_time=None):
        self.order_id = Order.next_order_id
        Order.next_order_id += 1
        self.user = user
        self.restaurant = restaurant
        self.items = items
        self.order_type = order_type
        self.scheduled_time = scheduled_time

    def get_order_id(self):
        return self.order_id

    def get_user(self):
        return self.user

    def get_restaurant(self):
        return self.restaurant

    def get_items(self):
        return self.items

    def get_total(self):
        return sum(item.get_price() for item in self.items)

    def get_type(self):
        return self.order_type

    def get_scheduled(self):
        return self.scheduled_time

File: Lecture 11/Java Code/Tomato/models/Order.js
export class Order {
    static nextOrderId = 0;

    constructor() {
        if (this.constructor === Order) {
            // Abstract class simulation
            // throw new Error("Cannot instantiate abstract class Order");
        }
        this.user = null;
        this.restaurant = null;
        this.paymentStrategy = null;
        this.total = 0.0;
        this.scheduled = "";
        this.orderId = ++Order.nextOrderId;
        this.items = [];
    }

    processPayment() {
        if (this.paymentStrategy !== null) {
            this.paymentStrategy.pay(this.total);
            return true;
        } else {
            console.log("Please choose a payment mode first");
            return false;
        }
    }

    getType() {
        throw new Error("Method 'getType()' must be implemented.");
    }

    getOrderId() {
        return this.orderId;
    }

    setUser(u) {
        this.user = u;
    }

    getUser() {
        return this.user;
    }

    setRestaurant(r) {
        this.restaurant = r;
    }

    getRestaurant() {
        return this.restaurant;
    }

    setItems(its) {
        this.items = its;
        this.total = 0;
        for (const i of this.items) {
            this.total += i.getPrice();
        }
    }

    getItems() {
        return this.items;
    }

    setPaymentStrategy(p) {
        this.paymentStrategy = p;
    }

    setScheduled(s) {
        this.scheduled = s;
    }

    getScheduled() {
        return this.scheduled;
    }

    getTotal() {
        return this.total;
    }

    setTotal(total) {
        this.total = total;
    }
}
